clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2655,6 +2960,87 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     addFieldInit(FD, Ty, ParamRef);
   }
 
+  Expr *addDerivedToBaseCastExpr(const CXXRecordDecl *RD,
+                                 const CXXBaseSpecifier &BS,
+                                 Expr *LocalCloneRef) {
+    CXXCastPath BasePath;
+    QualType DerivedTy(RD->getTypeForDecl(), 0);
+    QualType BaseTy = BS.getType();
+    SemaRef.CheckDerivedToBaseConversion(DerivedTy, BaseTy, KernelCallerSrcLoc,
+                                         SourceRange(), &BasePath,
+                                         /*IgnoreBaseAccess*/ true);
+    auto Cast = ImplicitCastExpr::Create(
+        SemaRef.Context, SemaRef.Context.getPointerType(BaseTy),
+        CK_DerivedToBase, LocalCloneRef,
+        /* CXXCastPath=*/&BasePath, VK_LValue, FPOptionsOverride());
+    return Cast;
+  }
+
+  Expr *getAddressOf(Expr *E) {
+    return UnaryOperator::Create(SemaRef.Context, E, UO_AddrOf,
+                                 SemaRef.Context.getPointerType(E->getType()),
+                                 VK_PRValue, OK_Ordinary, KernelCallerSrcLoc,
+                                 false, SemaRef.CurFPFeatureOverrides());
+  }
+
+  Expr *buildMemCpyCall(Expr *From, Expr *To, QualType T) {
```

Comment:
I remember in the first versions of the patch some Sema routine was added and re-used here. Is there any reason to make SemaSYCL-specific function?